packages/slate/src/transforms/text.ts

```diff
@@ -187,9 +187,13 @@ export const TextTransforms: TextTransforms = {
         }
       }
 
-      for (const pathRef of pathRefs) {
-        const path = pathRef.unref()!
-        Transforms.removeNodes(editor, { at: path, voids })
+      const middlePaths = pathRefs
```

Comment:
I think the idea of reversing the order of operation to remove without affecting `pathRefs` is a good one.

I'm curious why are we first creating an array using array functions and then using `for of`. What are the benefits of what you did here versus:

```ts
pathRefs
  .reverse()
  .map(r => r.unref())
  .filter((r): r is Path => r !== null) // note i would compare with null as per the unref interface
  .forEach(p => Transforms.removeNodes(editor, { at: p, voids })
```

Also, would it be possible to not use `pathRefs` at all given the fact that all `removeNodes` operations would work and the code before `removeNodes` doesn't seem to alter the tree. It is possible that not using `pathRefs` might be asking for trouble because something might get in between `matches` and `removeNodes` but seems like a possible performance gain (and performance gains are usually... less pretty).

Reply:
1. removed unnecessary `for...of`
2. Yes, pathRef can be removed, but currently there is only one remove_text, which has tiny impact on performance. Secondly, as you describe, using pathRef correctly expresses the relationship  between matches and removeNodes。

I think the importance is to let developers pay attention to the performance impact of `path/point/range ref` on `apply` Op. The mechanism based on location ref is prone to O(n2) operations, which will affect the experience of edit large fragment.